File: 필수문항/117_오목.py
import sys

input = sys.stdin.readline

graph = [list(map(int, input().split())) for _ in range(19)]
direction = [(-1, -1), (-1, 0), (-1, 1), (0, -1)]


def valid(x, y, num):
    global graph
    return 0 <= x < 19 and 0 <= y < 19 and graph[x][y] == num


def check(x, y, dx, dy):
    global graph
    count = 1
    num = graph[x][y]
    while valid(x + dx, y + dy, num):
        x += dx
        y += dy
        count += 1
    return count


for x in range(19):
    for y in range(19):
        if graph[x][y] != 0:
            for dx, dy in direction:
                count = 1

                if valid(x + dx, y + dy, graph[x][y]):
                    count += check(x + dx, y + dy, dx, dy)

                if valid(x - dx, y - dy, graph[x][y]):
                    count += check(x - dx, y - dy, -dx, -dy)

                if count == 5:
                    if dy == 1:
                        dx = -dx
                        dy = -dy
                    elif dy == 0 and dx == 1:
                        dx = -dx
                    num = graph[x][y]
                    while valid(x + dx, y + dy, num):
                        x += dx
                        y += dy
                    print(num)
                    print(x + 1, y + 1)
                    exit()

print(0)

File: 필수문항/117_오목.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.Arrays;

public class Main {
    public static int[][] graph = new int[19][19];
    public static int[][] direction = { { -1, -1 }, { -1, 0 }, { -1, 1 }, { 0, -1 } };

    public static boolean valid(int x, int y, int num) {
        return 0 <= x && x < 19 && 0 <= y && y < 19 && graph[x][y] == num;
    }

    public static int check(int x, int y, int dx, int dy) {
        int count = 1;
        int num = graph[x][y];
        while (valid(x + dx, y + dy, num)) {
            x += dx;
            y += dy;
            count++;
        }
        return count;
    }

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

        for (int i = 0; i < 19; i++) {
            graph[i] = Arrays.stream(br.readLine().split(" ")).mapToInt(Integer::parseInt).toArray();
        }
        for (int x = 0; x < 19; x++) {
            for (int y = 0; y < 19; y++) {
                if (graph[x][y] != 0) {
                    for (int[] d : direction) {
                        int dx = d[0];
                        int dy = d[1];
                        int count = 1;

                        if (valid(x + dx, y + dy, graph[x][y])) {
                            count += check(x + dx, y + dy, dx, dy);
                        }

                        if (valid(x - dx, y - dy, graph[x][y])) {
                            count += check(x - dx, y - dy, -dx, -dy);
                        }

                        if (count == 5) {
                            if (dy == 1) {
                                dx = -dx;
                                dy = -dy;
                            } else if (dy == 0 && dx == 1) {
                                dx = -dx;
                            }
                            int num = graph[x][y];
                            while (valid(x + dx, y + dy, num)) {
                                y += dy;
                                x += dx;
                            }
                            System.out.println(num);
                            System.out.printf("%d %d", x + 1, y + 1);
                            return;
                        }
                    }
                }
            }
        }
        System.out.println(0);

    }
}
